src/lib.rs

```diff
@@ -572,24 +566,44 @@ fn derive_where_internal(attr: TokenStream, item: TokenStream) -> Result<TokenSt
     } = syn::parse2(item)?;
 
     // Every trait needs a separate implementation.
-    for trait_ in derive_where.traits() {
+    for trait_ in derive_where.traits {
         let body = trait_.generate_body(&ident, &data)?;
-        let trait_ = trait_.type_();
+        let trait_ = trait_.path();
 
         // Build necessary generics to construct the implementation item.
-        let (impl_generics, type_generics, where_clause) = generics.split_for_impl();
-        // TODO: don't convert to `TokenStream`, but actually properly insert bounds.
-        let mut where_clause = where_clause.map(|where_clause| where_clause.into_token_stream());
+        let (impl_generics, type_generics, mut where_clause) = generics.split_for_impl();
+
+        let default_where = &WhereClause {
+            where_token: Where::default(),
+            predicates: Punctuated::default(),
+        };
+
+        let mut where_clause = where_clause.get_or_insert(default_where).clone();
 
         // If there are any bounds, insert them into the `where` clause.
-        if let DeriveWhere::WithGenerics(generics, _) = &derive_where {
-            // If there is no `where` clause, make one.
-            let where_clause = where_clause.get_or_insert(quote! { where });
-            let bounds = generics.iter().map(|generic| match generic {
-                Generic::CoustomBound(type_bound) => quote! {#type_bound},
-                Generic::NoBound(type_) => quote! {#type_: #trait_},
-            });
-            *where_clause = quote! { #where_clause #(#bounds),* };
+        for generic in &derive_where.generics {
+            where_clause
+                .predicates
+                .push(WherePredicate::Type(match generic {
+                    Generic::CoustomBound(type_bound) => type_bound.clone(),
+                    Generic::NoBound(type_) => {
+                        let mut bounds = Punctuated::default();
+
+                        bounds.push(TypeParamBound::Trait(TraitBound {
+                            paren_token: None,
+                            modifier: syn::TraitBoundModifier::None,
+                            lifetimes: None,
+                            path: trait_.clone(),
+                        }));
+
+                        PredicateType {
+                            lifetimes: None,
+                            bounded_ty: type_.clone(),
+                            colon_token: Colon::default(),
+                            bounds,
+                        }
```

Comment:
```suggestion
syn::parse2(quote! { type_: trait_})
```
Could also work here, right?

Reply:
should we be reparsing our output just to output it a few lines later again?

Reply:
No agreed, this makes no sense, maybe put a comment on top saying that this whole code block is basically `syn::parse2(quote! { type_: trait_})`.

Alternatively we could just revert this whole thing, I'm in favor of leaving it though, it's better then parsing stuff.

Reply:
I would expect there not being a performance problem with the previous version.
This will just be returned as a token stream, so there should be no parsing in the quote happening.

I do see your point about accuracy!

Reply:
It could make sense to move some of these in their own functions to make this more readable. But I would leave it for now, and come back later to it.

Reply:
Sounds good.